runtime/vm/ObjectMonitor.cpp

```diff
@@ -283,31 +310,119 @@ objectMonitorEnterNonBlocking(J9VMThread *currentThread, j9object_t object)
+			 */
+			if (0 == (lock & OBJECT_HEADER_LOCK_LEARNING)) {
+				/* Flat and Reserved state case */
+				J9_STORE_LOCKWORD(currentThread, lwEA, incremented);
+			} else {
+				/*
+				 * Learning state case
+				 * The Learning Counter is only used in the Learning state.
+				 * It starts at 0 the first time an object is locked. If the same thread locks the object again, the counter is incremented.
+				 * When the Learning Counter reaches reservedTransitionThreshold, the object transitions from Learning to Reserved.
+				 * reservedTransitionThreshold is being checked against a 2 bit value so values of 4 or higher will prevent transitions to Reserved.
+				 */
+				U_32 reservedTransitionThreshold = currentThread->javaVM->reservedTransitionThreshold;
+				bool reservedTransition = false;
+				if (((lock & OBJECT_HEADER_LOCK_LEARNING_LC_MASK) >> OBJECT_HEADER_LOCK_LEARNING_LC_OFFSET) >= reservedTransitionThreshold) {
+					/*
+					 * Clears the Learning bit and sets the Reserved bit to change the state to Reserved.
+					 * Clears the bits under OBJECT_HEADER_LOCK_RECURSION_MASK then fills in the new RC.
+					 * This is done to reclaim the bits used for the Learning Counter that are no longer used under Reserved.
+					 */
+					incremented = (incremented & ~((j9objectmonitor_t)OBJECT_HEADER_LOCK_LEARNING | (j9objectmonitor_t)OBJECT_HEADER_LOCK_RECURSION_MASK)) |
+					              (J9_FLATLOCK_COUNT(incremented) << OBJECT_HEADER_LOCK_V2_RECURSION_OFFSET) |
+					              OBJECT_HEADER_LOCK_RESERVED;
+					reservedTransition = true;
+				} else {
+					/* LC field overflow check. Don't increment if it will overflow the counter. */
+					if ((lock & OBJECT_HEADER_LOCK_LEARNING_LC_MASK) != OBJECT_HEADER_LOCK_LEARNING_LC_MASK) {
+						incremented = incremented + OBJECT_HEADER_LOCK_LEARNING_FIRST_LC_BIT;
+					}
+				}
+
+				if (lock == VM_ObjectMonitor::compareAndSwapLockword(currentThread, lwEA, lock, incremented, false)) {
+					if (0 == J9_FLATLOCK_COUNT(lock)) {
+						VM_AtomicSupport::monitorEnterBarrier();
+					}
+					if (reservedTransition) {
+						/* Transition from Learning to Reserved occurred so the Reserved Counter in the object's J9Class is incremented by 1. */
+						VM_ObjectMonitor::incrementReservedCounter(J9OBJECT_CLAZZ(currentThread, object));
+					}
+				} else {
+					/*
+					 * Another thread has attempted to get the lock and atomically changed the state to Flat.
+					 * Start over and read the new lockword. The lock can not go back to Learning so this will
+					 * only happen once.
+					 */
+					goto restart;
+				}
+			}
 			/* no barrier is required in the recursive case */
 		} else {
 			/* check to see if object is unlocked (JIT did not do initial inline sequence due to insufficient static type info) */
-			if ((0 == (lock & ~OBJECT_HEADER_LOCK_RESERVED)) &&
+			if ((0 == (lock & ~(j9objectmonitor_t)OBJECT_HEADER_LOCK_RESERVED & ~(j9objectmonitor_t)OBJECT_HEADER_LOCK_LEARNING)) &&
```

Comment:
The extra cast obscures what is happening here:
```
if ((0 == (lock & ~(j9objectmonitor_t)OBJECT_HEADER_LOCK_RESERVED & ~(j9objectmonitor_t)OBJECT_HEADER_LOCK_LEARNING)) &&
```
it could be rewritten as:
```
if ((0 == (lock & ~(j9objectmonitor_t)(OBJECT_HEADER_LOCK_RESERVED & OBJECT_HEADER_LOCK_LEARNING))) &&
```

Reply:
I can change that. The check is supposed to check if every bit in the lockword except the reserved and learning bits are 0 so it will actually be `OBJECT_HEADER_LOCK_RESERVED | OBJECT_HEADER_LOCK_LEARNING` before the cast.

Reply:
Even better would be to make use of `J9_ARE_NO_BITS_SET` like elsewhere in this file.

Reply:
That's a good idea. I just added that in.